devfile/devworkspace-operator
# Add support for postStart events in DevWorkspaces

### What does this PR do?
Adds basic support for postStart events in DevWorkspaces by utilizing Kubernetes [pod lifecycle hooks](https://kubernetes.io/docs/tasks/configure-pod-container/attach-handler-lifecycle-event/). Hooks allow for running commands after a container starts, with the caveats:

* There's no guarantee the container's entrypoint has run
* The command is re-run on every container start (i.e. if a container crashes the hook executes again)
* If the hook returns a non-zero exit code, container startup fails

This PR has some additional limitations:
* Only exec-type commands are supported (perhaps we could also support composite commands if they're all exec...)
* The `workingDir` field is supported via `cd <workingDir>` but other fields (e.g. `env`) are currently not
* Only one postStart event can be tied to each component, we don't join commands

The limitations above can be improved later if necessary.

### What issues does this PR fix or reference?
Closes https://github.com/devfile/devworkspace-operator/issues/629
Closes https://github.com/devfile/devworkspace-operator/issues/655

### Is it tested? How?
1. Create a DevWorkspace that specifies a postStart event:
    ```yaml
    cat <<EOF | kubectl apply -f -
    kind: DevWorkspace
    apiVersion: workspace.devfile.io/v1alpha2
    metadata:
      name: test-poststart
    spec:
      started: true
      template:
        projects:
          - name: web-nodejs-sample
            git:
              remotes:
                origin: "https://github.com/che-samples/web-nodejs-sample.git"
        components:
          - name: theia
            plugin:
              uri: https://che-plugin-registry-main.surge.sh/v3/plugins/eclipse/che-theia/next/devfile.yaml
              components:
                - name: theia-ide
                  container:
                    env:
                      - name: THEIA_HOST
                        value: 0.0.0.0
        commands:
          - id: say-hello
            exec:
              component: theia-ide
              commandLine: echo "Hello from postStart!" > postStart-event.log
              workingDir: ${PROJECTS_ROOT}
        events:
          postStart:
            - say-hello
    EOF
    ```
2. Wait for workspace to enter running state
3. Check that `postStart-event.log` file is created:
    ```bash
    WKSP_ID=$(kubectl get dw test-poststart -o json | jq -r '.status.devworkspaceId')
    kubectl exec deploy/${WKSP_ID} -c theia-ide -- cat '/projects/postStart-event.log'
    ```
    should output `Hello from postStart!`

### PR Checklist
